ga4gh/gff3Parser.py

```diff
@@ -0,0 +1,321 @@
+        self.score = score
+        self.strand = strand
+        self.frame = frame
+        self.attributes = copy.deepcopy(attributes)
+        self.gff3Set = gff3Set
+        self.lineNumber = lineNumber
+        self.parents = set()
+        self.children = set()
+
+    @staticmethod
+    def __dotIfNone(val):
+        return val if val is not None else '.'
+
+    def __attributeStr(self, name):
+        """
+        Return name=value for a single attribute
+        """
+        return "{}={}".format(
+            _encodeAttr(name),
+            ",".join([_encodeAttr(v) for v in self.attributes[name]]))
+
+    def __attributeStrs(self):
+        """
+        Return name=value, semi-colon-separated string for attributes,
+        including url-style quoting
+        """
+        return ";".join([self.__attributeStr(name)
+                         for name in self.attributes.iterkeys()])
+
+    def __str__(self):
+        """
+        Return the object as a valid GFF3 record line.
+        """
+        return "\t".join([self.seqname, self.source, self.type,
+                          str(self.start), str(self.end),
+                          self.__dotIfNone(self.score),
+                          self.__dotIfNone(self.strand),
+                          self.__dotIfNone(self.frame),
+                          self.__attributeStrs()])
+
+    @property
+    def featureId(self):
+        """
+        ID attribute or None if records doesn't have an ID
+        """
+        featId = self.attributes.get("ID")
+        if featId is not None:
+            featId = featId[0]
+        return featId
+
+    def getRequiredAttribute(self, name):
+        "get a require attributes list of values, error if not found"
+        values = self.attributes.get(name)
+        if values is None:
+            raise GFF3Exception(
+                "required attribute not found: {}".format(name),
+                self.gff3Set.fileName, self.lineNumber)
+        return values
+
+
+class Gff3Set(object):
+    """
+    A set of GFF3 sequence annotations
+    """
+    def __init__(self, fileName=None):
+        self.fileName = fileName
+        self.roots = set()     # root nodes (those with out parents)
+        # index of features by id. GFF3 allows disjoint features with
+        # the same id.  None is used to store features without ids
+        self.byFeatureId = collections.defaultdict(list)
+
+    def add(self, feature):
+        """
+        Add a feature record
+        """
+        # None featureId allowed
+        self.byFeatureId[feature.featureId].append(feature)
+
+    def __linkFeature(self, feature):
+        """
+        Link a feature with it's parents.
+        """
+        parentIds = feature.attributes.get("Parent")
+        if parentIds is None:
+            self.roots.add(feature)
+        else:
+            for parentId in parentIds:
+                self.__linkToParent(feature, parentId)
+
+    def __linkToParent(self, feature, parentId):
```

Comment:
@jeromekelleher I am, again, not argument against the change.
Local convention is important, having dealt with code that had
little.  I just wanting to understand the rationale.

I would be concerned if `_' was there to indicate things can be
overridden rather than just communicating data hiding.

A tool that does better checking would save you time. I think
this is especially important for the 2 vs 3, since there are
subtle things.  I have looked at packages that tried to abstract
this away, and it didn't seem elegant enough to pursue.

Jerome Kelleher <notifications@github.com> writes:
> OK, well we need make a choice then I think:
> 
> 1) Use the code here and follow the local coding conventions;
> 2) Explicitly make your code an external dependency and import it from pypi.
> 
> I don't really care that much about private/protected methods, but uniformity
> of style is essential.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub*
>